<a href="https://colab.research.google.com/github/chandan-DataDude-gif/-MYDAILYWORK/blob/main/Customer_churn_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("/content/Churn_Modelling.csv")

In [3]:
df.head(5)

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [5]:
df.shape

(10000, 14)

In [6]:
df.describe()

,RowNumber,CustomerId,CreditScore,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
count,10000.00000,1.000000e+04,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000
mean,5000.50000,1.569094e+07,650.528800,38.921800,5.012800,76485.889288,1.530200,0.70550,0.515100,100090.239881,0.203700
std,2886.89568,7.193619e+04,96.653299,10.487806,2.892174,62397.405202,0.581654,0.45584,0.499797,57510.492818,0.402769
min,1.00000,1.556570e+07,350.000000,18.000000,0.000000,0.000000,1.000000,0.00000,0.000000,11.580000,0.000000
25%,2500.75000,1.562853e+07,584.000000,32.000000,3.000000,0.000000,1.000000,0.00000,0.000000,51002.110000,0.000000
50%,5000.50000,1.569074e+07,652.000000,37.000000,5.000000,97198.540000,1.000000,1.00000,1.000000,100193.915000,0.000000
75%,7500.25000,1.575323e+07,718.000000,44.000000,7.000000,127644.240000,2.000000,1.00000,1.000000,149388.247500,0.000000
max,10000.00000,1.581569e+07,850.000000,92.000000,10.000000,250898.090000,4.000000,1.00000,1.000000,199992.480000,1.000000


In [7]:
df.isna().sum()

,0
RowNumber,0
CustomerId,0
Surname,0
CreditScore,0
Geography,0
Gender,0
Age,0
Tenure,0
Balance,0
NumOfProducts,0


In [8]:
df = df.drop(["RowNumber","CustomerId","Surname"],axis = 1)

In [9]:
df.head(5)

,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


In [10]:
x = df.drop(["Exited"], axis = 1)
y = df["Exited"]

In [11]:
#splitting the data into train and testing

In [12]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(
    x,y,test_size = 0.2, random_state = 42
)

In [13]:
#Using the Logisticsregression model to check the accuraacy score and to know how many customer churn.

In [14]:
from sklearn.linear_model import LogisticRegression
import pandas as pd

x_train = pd.get_dummies(x_train, columns=['Geography', 'Gender'], drop_first=True)
x_test = pd.get_dummies(x_test, columns=['Geography', 'Gender'], drop_first=True)

model = LogisticRegression(max_iter=1000)
model.fit(x_train , y_train)
y_pred = model.predict(x_test)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [15]:
y_pred

array([0, 0, 0, ..., 0, 0, 0])

In [16]:
df["Exited"].value_counts()

,count
Exited,
0,7963
1,2037


In [17]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
print("accuracy : ", accuracy_score(y_test , y_pred))
print("confusion_matrix : ", confusion_matrix(y_test , y_pred))
print("classification_report : ", classification_report(y_test , y_pred))

accuracy :  0.8105
confusion_matrix :  [[1532   75]
 [ 304   89]]
classification_report :                precision    recall  f1-score   support

           0       0.83      0.95      0.89      1607
           1       0.54      0.23      0.32       393

    accuracy                           0.81      2000
   macro avg       0.69      0.59      0.60      2000
weighted avg       0.78      0.81      0.78      2000



In [18]:
#Using the randomforestclassifier for the better prediction and the output

In [19]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
rf = RandomForestClassifier(
    n_estimators = 100,
    class_weight = "balanced",
    random_state = 42
)
rf.fit(x_train , y_train)
y_pred = rf.predict(x_test)
print("accuracy : ", accuracy_score(y_test , y_pred))
print("classification_report : ", classification_report(y_test , y_pred))
print("confusion_matrix : ", confusion_matrix(y_test , y_pred))

accuracy :  0.869
classification_report :                precision    recall  f1-score   support

           0       0.88      0.97      0.92      1607
           1       0.79      0.46      0.58       393

    accuracy                           0.87      2000
   macro avg       0.83      0.71      0.75      2000
weighted avg       0.86      0.87      0.85      2000

confusion_matrix :  [[1559   48]
 [ 214  179]]


In [20]:
#Using the  gradientboosting for the better result.

In [37]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import roc_auc_score
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
gb = GradientBoostingClassifier(
    n_estimators = 200,
    random_state = 42
)
gb.fit(x_train , y_train)
y_pred = gb.predict(x_test)
y_prob = gb.predict_proba(x_test)[:,1]
y_pred_new = (y_prob > 0.30).astype(int)
print("accuracy_score : ", accuracy_score(y_test, y_pred_new))
print("classification_report : ", classification_report(y_test, y_pred_new))
print("confusion_matrix : ", confusion_matrix(y_test , y_pred_new))
print("Roc_Auc_score : ", roc_auc_score(y_test, y_prob))

accuracy_score :  0.8555
classification_report :                precision    recall  f1-score   support

           0       0.92      0.90      0.91      1607
           1       0.62      0.67      0.65       393

    accuracy                           0.86      2000
   macro avg       0.77      0.79      0.78      2000
weighted avg       0.86      0.86      0.86      2000

confusion_matrix :  [[1447  160]
 [ 129  264]]
Roc_Auc_score :  0.8695069756836742


#Overall cunclusion after using the logisticregression, Gradientboosting, RandomForestClassifier
#gradientboosting is better in terms of predicting the customer churn behavior than randomforest and logisticsregression.
#after all the model detected 49% cutomer who are churn and unable to detect the 51% customer who actually churn.
#after changing the threshold value from 0.5 to 0.30 model recall inhanced by 13% and the accuracy of the model is also very good.

#after using smote their is a very minimal optimisation can be seen becouse the gradientboosting is already a powerful model thats why the impact of the smote is minimal therefore i move forward without smote.

recall  = 67 % (means that model is detecting 67 churn customer out of 100 it can be fixed by changing the threshold value but the precision and accuracy can be change).
roc_auc_acore = 87 % (model is strong)